code/modules/admin/known_alts.dm

```diff
@@ -0,0 +1,186 @@
+			if (!ckey2)
+				return
+
+			ckey1 = ckey(ckey1)
+			ckey2 = ckey(ckey2)
+
+			var/datum/db_query/query_already_exists = SSdbcore.NewQuery({"
+				SELECT id FROM [format_table_name("known_alts")]
+				WHERE (ckey1 = :ckey1 AND ckey2 = :ckey2)
+				OR (ckey1 = :ckey2 AND ckey2 = :ckey1)
+			"}, list(
+				"ckey1" = ckey1,
+				"ckey2" = ckey2,
+			))
+
+			query_already_exists.warn_execute()
+
+			if (query_already_exists.last_error)
+				qdel(query_already_exists)
+				return
+
+			if (query_already_exists.NextRow())
+				alert(usr, "Those two are already in the list of known alts!")
+				return
+
+			qdel(query_already_exists)
+
+			var/datum/db_query/query_add_known_alt = SSdbcore.NewQuery({"
+				INSERT INTO [format_table_name("known_alts")] (ckey1, ckey2)
+				VALUES (:ckey1, :ckey2)
+			"}, list(
+				"ckey1" = ckey1,
+				"ckey2" = ckey2,
+			))
+
+			if (query_add_known_alt.warn_execute())
+				var/message = "[key_name(usr)] has added a new known alt connection between [ckey1] and [ckey2]."
+				message_admins(message)
+				log_admin_private(message)
+
+				cached_known_alts = null
+				load_known_alts()
+
+			qdel(query_add_known_alt)
+			show_panel(usr.client)
+
+			if (!is_banned_from(ckey2, "Server"))
+				var/ban_choice = alert("[ckey2] is not banned from the server. Do you want to open up the ban panel as well?",,"Yes", "No")
+				if (ban_choice == "Yes")
+					holder.ban_panel(ckey2, role = "Server", duration = BAN_PANEL_PERMANENT)
+		if ("delete")
+			var/id = text2num(href_list["id"])
+			if (!id)
+				log_admin_private("[key_name(usr)] tried to delete an invalid known alt ID: [href_list["id"]]")
+				return
+
+			var/datum/db_query/query_known_alt_info = SSdbcore.NewQuery({"
+				SELECT ckey1, ckey2 FROM [format_table_name("known_alts")]
+				WHERE id = :id
+			"}, list(
+				"id" = id,
+			))
+
+			if (!query_known_alt_info.warn_execute())
+				qdel(query_known_alt_info)
+				return
+
+			if (!query_known_alt_info.NextRow())
+				alert("Couldn't find the known alt with the ID [id]")
+				return
+
+			var/list/result = query_known_alt_info.item
+
+			if (alert("Are you sure you want to delete the alt connection between [result[1]] and [result[2]]?",,"Yes", "No") != "Yes")
+				return
+
+			var/datum/db_query/query_delete_known_alt = SSdbcore.NewQuery({"
+				DELETE FROM [format_table_name("known_alts")]
+				WHERE id = :id
+			"}, list(
+				"id" = id,
+			))
+
+			if (query_delete_known_alt.warn_execute())
+				var/message = "[key_name(usr)] has deleted the known alt connection between [result[1]] and [result[2]]."
+				message_admins(message)
+				log_admin_private(message)
+
+				cached_known_alts = null
+				load_known_alts()
+
+			qdel(query_delete_known_alt)
+			show_panel(usr.client)
+
+/// Returns the list of known alts, will return an empty list if the DB could not be connected to.
+/// This proc can block.
+/datum/known_alts/proc/load_known_alts()
+	if (!isnull(cached_known_alts))
+		return cached_known_alts
```

Comment:
Cached sql data has to be timed and have a ttl.

If the ttl is expired, but it fails to fetch the data, it should then return the stale cached data.

See the stickyban subsystem for further details.

https://github.com/tgstation/tgstation/blob/master/.github/guides/STANDARDS.md#sql:

>The ttl for data from the database is 10 seconds. You must have a compelling reason to store and reuse data for longer then this.